File: project2/WorldFrame.java
import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import java.util.*;
import java.io.*;

class WorldFrame extends JFrame {
    private static final int WIDTH = 40;
    private static final int HEIGHT = 20;
    private static final int CELL_SIZE = 30;
    
    private World world;
    private WorldPanel worldPanel;
    private JLabel statusLabel;
    private JLabel humanStatusLabel;
    private JLabel strengthLabel;
    
    public WorldFrame() {
        super("World Simulation - Oskar Putz ID: 203475");
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        
        world = new World();
        initializeWorld();
        
        setLayout(new BorderLayout());
        
        worldPanel = new WorldPanel();
        add(worldPanel, BorderLayout.CENTER);
        
        JPanel statusPanel = new JPanel(new GridLayout(3, 1));
        statusLabel = new JLabel("No collisions yet.");
        humanStatusLabel = new JLabel("Ability status: OFF");
        strengthLabel = new JLabel("Current strength: " + world.human.strength);
        statusPanel.add(humanStatusLabel);
        statusPanel.add(strengthLabel);
        statusPanel.add(statusLabel);
        add(statusPanel, BorderLayout.SOUTH);
        
        JPanel controlsPanel = new JPanel();
        JButton upButton = new JButton("Up");
        JButton downButton = new JButton("Down");
        JButton leftButton = new JButton("Left");
        JButton rightButton = new JButton("Right");
        JButton abilityButton = new JButton("Activate Ability");
        JButton nextTurnButton = new JButton("Next Turn");
        JButton saveButton = new JButton("Save");
        JButton loadButton = new JButton("Load");
        
        upButton.addActionListener(e -> moveHuman(0, -1));
        downButton.addActionListener(e -> moveHuman(0, 1));
        leftButton.addActionListener(e -> moveHuman(-1, 0));
        rightButton.addActionListener(e -> moveHuman(1, 0));
        abilityButton.addActionListener(e -> activateHumanAbility());
        nextTurnButton.addActionListener(e -> nextTurn());
        saveButton.addActionListener(e -> saveWorld());
        loadButton.addActionListener(e -> loadWorld());
        
        controlsPanel.add(upButton);
        controlsPanel.add(downButton);
        controlsPanel.add(leftButton);
        controlsPanel.add(rightButton);
        controlsPanel.add(abilityButton);
        controlsPanel.add(nextTurnButton);
        controlsPanel.add(saveButton);
        controlsPanel.add(loadButton);
        add(controlsPanel, BorderLayout.NORTH);
        
        // Set up keyboard shortcuts
        KeyStroke upKey = KeyStroke.getKeyStroke(KeyEvent.VK_UP, 0);
        KeyStroke downKey = KeyStroke.getKeyStroke(KeyEvent.VK_DOWN, 0);
        KeyStroke leftKey = KeyStroke.getKeyStroke(KeyEvent.VK_LEFT, 0);
        KeyStroke rightKey = KeyStroke.getKeyStroke(KeyEvent.VK_RIGHT, 0);
        KeyStroke pKey = KeyStroke.getKeyStroke(KeyEvent.VK_P, 0);
        KeyStroke spaceKey = KeyStroke.getKeyStroke(KeyEvent.VK_SPACE, 0);
        
        getRootPane().registerKeyboardAction(e -> moveHuman(0, -1), upKey, JComponent.WHEN_IN_FOCUSED_WINDOW);
        getRootPane().registerKeyboardAction(e -> moveHuman(0, 1), downKey, JComponent.WHEN_IN_FOCUSED_WINDOW);
        getRootPane().registerKeyboardAction(e -> moveHuman(-1, 0), leftKey, JComponent.WHEN_IN_FOCUSED_WINDOW);
        getRootPane().registerKeyboardAction(e -> moveHuman(1, 0), rightKey, JComponent.WHEN_IN_FOCUSED_WINDOW);
        getRootPane().registerKeyboardAction(e -> activateHumanAbility(), pKey, JComponent.WHEN_IN_FOCUSED_WINDOW);
        getRootPane().registerKeyboardAction(e -> nextTurn(), spaceKey, JComponent.WHEN_IN_FOCUSED_WINDOW);
        
        setSize(WIDTH * CELL_SIZE + 16, HEIGHT * CELL_SIZE + 150);
        setLocationRelativeTo(null);
        setVisible(true);
    }
        private void saveWorld() {
        JFileChooser fileChooser = new JFileChooser();
        fileChooser.setDialogTitle("Save World");
        
        if (fileChooser.showSaveDialog(this) == JFileChooser.APPROVE_OPTION) {
            File file = fileChooser.getSelectedFile();
            
            try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(file))) {
                oos.writeObject(world);
                statusLabel.setText("World saved successfully to " + file.getName());
            } catch (IOException e) {
                JOptionPane.showMessageDialog(this, 
                    "Error saving world: " + e.getMessage(), 
                    "Save Error", 
                    JOptionPane.ERROR_MESSAGE);
                e.printStackTrace();
            }
        }
    }
        private void loadWorld() {
        JFileChooser fileChooser = new JFileChooser();
        fileChooser.setDialogTitle("Load World");
        
        if (fileChooser.showOpenDialog(this) == JFileChooser.APPROVE_OPTION) {
            File file = fileChooser.getSelectedFile();
            
            try (ObjectInputStream ois = new ObjectInputStream(new FileInputStream(file))) {
                world = (World) ois.readObject();
                statusLabel.setText("World loaded successfully from " + file.getName());
                updateStatus();
                worldPanel.repaint();
            } catch (IOException | ClassNotFoundException e) {
                JOptionPane.showMessageDialog(this, 
                    "Error loading world: " + e.getMessage(), 
                    "Load Error", 
                    JOptionPane.ERROR_MESSAGE);
                e.printStackTrace();
            }
        }
    }
    
    private void moveHuman(int dx, int dy) {
        if (world.human != null) {
            world.human.setDirection(dx, dy);
            nextTurn();
        }
    }
    
    private void activateHumanAbility() {
        if (world.human != null) {
            world.human.activateAbility();
            updateStatus();
            worldPanel.repaint();
        }
    }
    
    private void nextTurn() {
        world.makeTurn();
        updateStatus();
        worldPanel.repaint();
    }
    
    private void updateStatus() {
        statusLabel.setText(world.lastCollisionMessage);
        humanStatusLabel.setText("Ability status: " + (world.human.abilityDuration > 0 ? "ON" : "OFF"));
        strengthLabel.setText("Current strength: " + world.human.strength);
    }
    
    private void initializeWorld() {
        Random random = new Random();
        world.human = new Human(0, 0, world);
        world.addOrganism(world.human);
        
        for (int i = 0; i < 2; i++) world.addOrganism(new Sheep(random.nextInt(WIDTH), random.nextInt(HEIGHT), world));
        for (int i = 0; i < 2; i++) world.addOrganism(new Wolf(random.nextInt(WIDTH), random.nextInt(HEIGHT), world));
        for (int i = 0; i < 2; i++) world.addOrganism(new Fox(random.nextInt(WIDTH), random.nextInt(HEIGHT), world));
        for (int i = 0; i < 2; i++) world.addOrganism(new Turtle(random.nextInt(WIDTH), random.nextInt(HEIGHT), world));
        for (int i = 0; i < 2; i++) world.addOrganism(new Antelope(random.nextInt(WIDTH), random.nextInt(HEIGHT), world));
        for (int i = 0; i < 2; i++) world.addOrganism(new Grass(random.nextInt(WIDTH), random.nextInt(HEIGHT), world));
        for (int i = 0; i < 2; i++) world.addOrganism(new SowThistle(random.nextInt(WIDTH), random.nextInt(HEIGHT), world));
        for (int i = 0; i < 2; i++) world.addOrganism(new Guarana(random.nextInt(WIDTH), random.nextInt(HEIGHT), world));
        for (int i = 0; i < 2; i++) world.addOrganism(new Belladonna(random.nextInt(WIDTH), random.nextInt(HEIGHT), world));
        for (int i = 0; i < 2; i++) world.addOrganism(new Hogweed(random.nextInt(WIDTH), random.nextInt(HEIGHT), world));
    }
    
    class WorldPanel extends JPanel {
        private static final HashMap<Character, Color> COLORS = new HashMap<>();
        
        static {
            COLORS.put('H', Color.BLUE);
            COLORS.put('S', Color.WHITE);
            COLORS.put('W', Color.DARK_GRAY);
            COLORS.put('F', Color.ORANGE);
            COLORS.put('T', Color.GREEN.darker().darker());
            COLORS.put('A', Color.YELLOW);
            COLORS.put('G', Color.GREEN);
            COLORS.put('s', Color.GREEN.darker());
            COLORS.put('g', Color.RED.darker());
            COLORS.put('b', Color.MAGENTA);
            COLORS.put('h', Color.RED);
        }
        
        @Override
        protected void paintComponent(Graphics g) {
            super.paintComponent(g);
            int width2 = 40;
            int height2 = 20;
            // Draw grid
            g.setColor(Color.LIGHT_GRAY);
            for (int x = 0; x <= width2; x++) {
                g.drawLine(x * CELL_SIZE, 0, x * CELL_SIZE, height2 * CELL_SIZE);
            }
            for (int y = 0; y <= height2; y++) {
                g.drawLine(0, y * CELL_SIZE, width2 * CELL_SIZE, y * CELL_SIZE);
            }
            
            // Draw organisms
            for (int y = 0; y < height2; y++) {
                for (int x = 0; x < width2; x++) {
                    if (world.grid[y][x] != null) {
                        char symbol = world.grid[y][x].draw();
                        g.setColor(COLORS.getOrDefault(symbol, Color.BLACK));
                        g.fillRect(x * CELL_SIZE + 1, y * CELL_SIZE + 1, CELL_SIZE - 1, CELL_SIZE - 1);
                        g.setColor(Color.BLACK);
                        g.drawString(String.valueOf(symbol), x * CELL_SIZE + CELL_SIZE/2 - 4, y * CELL_SIZE + CELL_SIZE/2 + 4);
                    }
                }
            }
        }
    }
}

File: project3/world_frame.py
import tkinter as tk
from tkinter import filedialog, messagebox
import pickle
import random
from world import World
from wolf import Wolf
from sheep import Sheep
from fox import Fox
from turtler import Turtler
from antelope import Antelope
from grass import Grass
from sowthistle import SowThistle
from guarana import Guarana
from belladonna import Belladonna
from hogweed import Hogweed
from human import Human
from cybersheep import CyberSheep

class WorldFrame(tk.Tk):
    WIDTH = 40
    HEIGHT = 20
    CELL_SIZE = 30
    
    def __init__(self):
        super().__init__()
        self.title("World Simulation - Oskar Putz")
        self.protocol("WM_DELETE_WINDOW", self.on_close)
        
        self.world = World()
        self.initialize_world()
        
        # Main frame
        main_frame = tk.Frame(self)
        main_frame.pack(fill=tk.BOTH, expand=True)
        
        # Control panel
        control_panel = tk.Frame(main_frame)
        control_panel.pack(side=tk.TOP, fill=tk.X)
        
        up_button = tk.Button(control_panel, text="Up", command=lambda: self.move_human(0, -1))
        down_button = tk.Button(control_panel, text="Down", command=lambda: self.move_human(0, 1))
        left_button = tk.Button(control_panel, text="Left", command=lambda: self.move_human(-1, 0))
        right_button = tk.Button(control_panel, text="Right", command=lambda: self.move_human(1, 0))
        ability_button = tk.Button(control_panel, text="Activate Ability", command=self.activate_human_ability)
        next_turn_button = tk.Button(control_panel, text="Next Turn", command=self.next_turn)
        save_button = tk.Button(control_panel, text="Save", command=self.save_world)
        load_button = tk.Button(control_panel, text="Load", command=self.load_world)
        
        up_button.pack(side=tk.LEFT, padx=5, pady=5)
        down_button.pack(side=tk.LEFT, padx=5, pady=5)
        left_button.pack(side=tk.LEFT, padx=5, pady=5)
        right_button.pack(side=tk.LEFT, padx=5, pady=5)
        ability_button.pack(side=tk.LEFT, padx=5, pady=5)
        next_turn_button.pack(side=tk.LEFT, padx=5, pady=5)
        save_button.pack(side=tk.LEFT, padx=5, pady=5)
        load_button.pack(side=tk.LEFT, padx=5, pady=5)
        
        # Canvas for world display
        self.canvas = tk.Canvas(main_frame, width=self.WIDTH * self.CELL_SIZE, 
                              height=self.HEIGHT * self.CELL_SIZE, bg="white")
        self.canvas.pack(side=tk.TOP, fill=tk.BOTH, expand=True)
        
        # Status bar
        status_panel = tk.Frame(main_frame)
        status_panel.pack(side=tk.BOTTOM, fill=tk.X)
        
        self.human_status_label = tk.Label(status_panel, text="Ability status: OFF")
        self.strength_label = tk.Label(status_panel, text=f"Current strength: {self.world.human.strength}")
        self.status_label = tk.Label(status_panel, text="No collisions yet.")
        
        self.human_status_label.pack(side=tk.TOP, anchor=tk.W, padx=5)
        self.strength_label.pack(side=tk.TOP, anchor=tk.W, padx=5)
        self.status_label.pack(side=tk.TOP, anchor=tk.W, padx=5)
        
        # Key bindings
        self.bind("<Up>", lambda e: self.move_human(0, -1))
        self.bind("<Down>", lambda e: self.move_human(0, 1))
        self.bind("<Left>", lambda e: self.move_human(-1, 0))
        self.bind("<Right>", lambda e: self.move_human(1, 0))
        self.bind("p", lambda e: self.activate_human_ability())
        self.bind("<space>", lambda e: self.next_turn())
        self.bind("q", lambda e: self.on_close())
        
        # Draw the initial world
        self.draw_world()
        
        # Set window size and position
        window_width = self.WIDTH * self.CELL_SIZE + 16
        window_height = self.HEIGHT * self.CELL_SIZE + 150
        screen_width = self.winfo_screenwidth()
        screen_height = self.winfo_screenheight()
        center_x = int(screen_width/2 - window_width/2)
        center_y = int(screen_height/2 - window_height/2)
        self.geometry(f'{window_width}x{window_height}+{center_x}+{center_y}')
    
    def initialize_world(self):
        # Add human
        human = Human(0, 0, self.world)
        self.world.human = human
        self.world.add_organism(human)
        
        # Add animals
        self.add_organisms(Wolf, 2)
        self.add_organisms(Sheep, 4)
        self.add_organisms(Fox, 2)
        self.add_organisms(Turtler, 2)
        self.add_organisms(Antelope, 2)
        self.add_organisms(CyberSheep, 2)
        
        # Add plants
        self.add_organisms(Grass, 4)
        self.add_organisms(SowThistle, 2)
        self.add_organisms(Guarana, 2)
        self.add_organisms(Belladonna, 2)
        self.add_organisms(Hogweed, 1)
    
    def add_organisms(self, organism_class, count):
        for _ in range(count):
            while True:
                x = random.randint(0, self.WIDTH - 1)
                y = random.randint(0, self.HEIGHT - 1)
                if self.world.grid[y][x] is None:
                    self.world.add_organism(organism_class(x, y, self.world))
                    break
    
    def draw_world(self):
        self.canvas.delete("all")
        
        # Draw grid lines
        for i in range(self.WIDTH + 1):
            self.canvas.create_line(i * self.CELL_SIZE, 0, 
                                  i * self.CELL_SIZE, self.HEIGHT * self.CELL_SIZE, 
                                  fill="gray")
        
        for i in range(self.HEIGHT + 1):
            self.canvas.create_line(0, i * self.CELL_SIZE, 
                                  self.WIDTH * self.CELL_SIZE, i * self.CELL_SIZE, 
                                  fill="gray")
        
        # Draw organisms
        colors = {
            'H': "blue",     # Human
            'W': "gray",     # Wolf
            'S': "white",    # Sheep
            'C': "cyan",     # CyberSheep
            'F': "orange",   # Fox
            'T': "green",    # Turtle
            'A': "brown",    # Antelope
            'G': "light green",  # Grass
            's': "yellow",   # SowThistle
            'g': "red",      # Guarana
            'b': "purple",   # Belladonna
            'h': "black"     # Hogweed
        }
        
        for y in range(self.HEIGHT):
            for x in range(self.WIDTH):
                organism = self.world.grid[y][x]
                if organism is not None:
                    char = organism.draw()
                    color = colors.get(char, "black")
                    
                    # Draw filled rectangle
                    self.canvas.create_rectangle(
                        x * self.CELL_SIZE + 1, 
                        y * self.CELL_SIZE + 1, 
                        (x + 1) * self.CELL_SIZE - 1, 
                        (y + 1) * self.CELL_SIZE - 1, 
                        fill=color
                    )
                    
                    # Draw character
                    self.canvas.create_text(
                        x * self.CELL_SIZE + self.CELL_SIZE // 2,
                        y * self.CELL_SIZE + self.CELL_SIZE // 2,
                        text=char,
                        fill="white" if color in ["blue", "black", "purple"] else "black"
                    )
        
        # Update status labels
        ability_status = "ON" if self.world.human.ability_duration > 0 else "OFF"
        cooldown = self.world.human.ability_cooldown
        
        self.human_status_label.config(
            text=f"Ability status: {ability_status}" + 
                 (f" (Cooldown: {cooldown})" if cooldown > 0 else "")
        )
        self.strength_label.config(text=f"Current strength: {self.world.human.strength}")
        self.status_label.config(text=self.world.last_collision_message)
    
    def move_human(self, dx, dy):
        if self.world.human is not None and self.world.human.is_alive():
            self.world.human.set_direction(dx, dy)
            self.next_turn()
    
    def activate_human_ability(self):
        if self.world.human is not None and self.world.human.is_alive():
            self.world.human.activate_ability()
            self.draw_world()
    
    def next_turn(self):
        self.world.make_turn()
        self.draw_world()
    
    def save_world(self):
        file_path = filedialog.asksaveasfilename(
            defaultextension=".pkl",
            filetypes=[("Pickle files", "*.pkl"), ("All files", "*.*")]
        )
        if file_path:
            try:
                with open(file_path, 'wb') as file:
                    pickle.dump(self.world, file)
                self.status_label.config(text=f"World saved successfully to {file_path}")
            except Exception as e:
                messagebox.showerror("Save Error", f"Error saving world: {str(e)}")
    
    def load_world(self):
        file_path = filedialog.askopenfilename(
            defaultextension=".pkl",
            filetypes=[("Pickle files", "*.pkl"), ("All files", "*.*")]
        )
        if file_path:
            try:
                with open(file_path, 'rb') as file:
                    self.world = pickle.load(file)
                self.draw_world()
                self.status_label.config(text=f"World loaded successfully from {file_path}")
            except Exception as e:
                messagebox.showerror("Load Error", f"Error loading world: {str(e)}")
    
    def on_close(self):
        if messagebox.askokcancel("Quit", "Do you want to quit?"):
            self.destroy()
